### PR TITLE
[WIP] Implemented part of core API

### LINKED ISSUE #8300: [GDNative] implement rest of the core API
There are a few core types which don't have all important methods exposed to the C API.

The most "dramatic" ones are

~~[godot_image](https://github.com/godotengine/godot/blob/master/modules/dlscript/godot/godot_image.h): I only implemented basic contructors, height and width, load and save. It would be cool to have at least all methods that are [exposed to varcall](http://docs.godotengine.org/en/stable/classes/class_image.html) implemented here. (if you want to go further and include functions only available to C++ then that's cool as well).~~

[godot_string](https://github.com/godotengine/godot/blob/master/modules/gdnative/godot/godot_string.h): Basically only has a constructor, a way to get the data back out of it, equality and concatenation. The [C++ class](https://github.com/godotengine/godot/blob/master/core/ustring.h) has a lot more functions that are also very important, (like `String::num()`). Also a function to get a `char *` would be nice, using `wchar_t *` can be a pain.

------------

~~Other types have missing C functions as well, but those functions can be implemented in the language binding much easier than string and image.~~

 * ~~[godot_vector3](https://github.com/godotengine/godot/blob/master/modules/dlscript/godot/godot_vector3.h)~~
 * ~~[godot_vector2](https://github.com/godotengine/godot/blob/master/modules/dlscript/godot/godot_vector2.h)~~
 * ~~[godot_transform2d](https://github.com/godotengine/godot/blob/master/modules/dlscript/godot/godot_transform2d.h)~~

~~Maybe there are more, but these are the ones I can think of right now. Probably all of the C wrapped core types lack functionality. Feel free to wrap the C++ classes as much as possible.~~

------------

~~Also, there's no equivalent of GDScript's `print()` or C++'s `print_line()`. So currently it's not possible to write to the in-editor-console using DLScript. A `godot_print()` function that does that would be cool. Such a function would live in [`godot.h/.c`](https://github.com/godotengine/godot/blob/master/modules/dlscript/godot.h)~~

------------

# Update

only really `string` methods are missing.
The rest is more or less complete, mostly thanks to @touilleMan  and also @sheepandshepherd.

Also @cynicaldevil for the efforts of wrapping Image, but that was removed recently, so your changes are obsolete 🙁 

If there are any questions, message me on IRC or Discord

### PR DESCRIPTION
**Pull request by nikhilshagri**
cc @karroffel 
Fixes #8300
I haven't tried to build it yet, just pushed in the state it is right now.

### CODE REVIEW

File: modules/gdnative/godot/godot_image.cpp [hunk 1]
```diff
@@ -105,6 +105,43 @@ int GDAPI godot_image_get_mipmap_count(const godot_image *p_img) {
 	return img->get_mipmap_count();
 }
 
+void GDAPI godot_image_blit_rect(godot_image *p_img, godot_image *src) {
+
+}
+
+void GDAPI godot_image_compressed(godot_image *p_img, int p_mode) {
+	Image *img = (Image *)p_img;
+	memnew_placement_custom(img, Image, img->compressed(p_mode));
```
Comment (karroffel) [inline modules/gdnative/godot/godot_image.cpp]: `memnew_placement_custom` is only used to *construct* the objects. If you just want to use them you can do

```cpp
Image *img = (Image *)p_img;
img->compressed(p_mode);
```

Comment (karroffel) [inline modules/gdnative/godot/godot_image.cpp]: For example take a look at the `has_mipmap` function.

File: modules/gdnative/godot/godot_image.h [hunk 2]
```diff
@@ -111,6 +112,14 @@ int GDAPI godot_image_get_height(const godot_image *p_img);
 godot_bool GDAPI godot_image_has_mipmaps(const godot_image *p_img);
 int GDAPI godot_image_get_mipmap_count(const godot_image *p_img);
 
+void GDAPI godot_image_blit_rect(godot_image *p_img, const godot_image *src, const godot_rect2 *p_src_rect, );
+void GDAPI godot_image_compressed(godot_image *p_img, int p_mode);
+void GDAPI godot_image_decompressed(const godot_image *p_img);
```
Comment (karroffel) [inline modules/gdnative/godot/godot_image.h]: This and `godot_image_compressed` should return a new image.

**Comment by cynicaldevil** [comment]
@karroffel I've updated the PR, and it builds ok now. Let me know if functions need to be added/updated.

File: modules/gdnative/godot/godot_image.cpp [hunk 3]
```diff
@@ -105,6 +105,46 @@ int GDAPI godot_image_get_mipmap_count(const godot_image *p_img) {
...
+void GDAPI godot_image_get_rect(const godot_image *p_img, godot_rect2 *p_rect) {
+	Image *img = (Image *)p_img;
+	Rect2 *cpp_rect2 = (Rect2 *)p_rect;
+	img->get_rect(*cpp_rect2);
+}
+
+godot_rect2 GDAPI godot_image_get_used_rect(const godot_image *p_img) {
+	Image *img = (Image *)p_img;
+	Rect2 cpp_rect2 = img->get_used_rect();
+	godot_rect2 *rect = (godot_rect2 *) &cpp_rect2;
```
Comment (bojidar-bg) [inline modules/gdnative/godot/godot_image.cpp]: Should be :
```cpp
godot_rect2 *rect = (godot_rect2 *)&cpp_rect2;
//---------------------------------^ No space here
```
According to styleguide

**Comment by cynicaldevil** [comment]
@bojidar-bg Sorry for the long delay, I had my end terms and some other stuff to attend to.I've updated the PR now.

File: modules/gdnative/godot/godot_image.cpp [hunk 4]
```diff
@@ -105,6 +105,46 @@ int GDAPI godot_image_get_mipmap_count(const godot_image *p_img) {
...
+	Image *src_img = (Image *)src;
+	Rect2 *cpp_src_rect = (Rect2 *)src_rect;
+	Vector2 *cpp_dest = (Vector2 *)dest;
+	img->blit_rect(*src_img, *cpp_src_rect, *cpp_dest);
+}
+
+godot_image GDAPI godot_image_compressed(godot_image *p_img, int p_mode) {
+	Image *img = (Image *)p_img;
+	Image cpp_img = img->compressed(p_mode);
+	godot_image *compressed_img = (godot_image *)&cpp_img;
```
Comment (karroffel) [inline modules/gdnative/godot/godot_image.cpp]: careful with this, it doesn't work. The destructor will be called at the end of the scope, taking a pointer doesn't change that.

So you need to create an object with non local lifetime using `memnew_placement` and then assign this new image to it.

This simple "pointer casting" technique doesn't work with objects that have a destructor.

**Review by karroffel — CHANGES_REQUESTED** [review]

File: modules/gdnative/godot/godot_image.cpp (hunk 4 above)
Comment (cynicaldevil) [inline modules/gdnative/godot/godot_image.cpp]: Soemthing like this?
```
godot_image GDAPI godot_image_compressed(godot_image *p_img, int p_mode) {
	Image *img = (Image *)p_img;
	godot_image compressed_img;
	Image *cpp_img = (Image *)&compressed_img;
	memnew_placement(cpp_img, Image);
	cpp_img = img->compressed(p_mode);
	return compressed_img;
}
```

Comment (karroffel) [inline modules/gdnative/godot/godot_image.cpp]: `*cpp_img = img->compressed(p_mode);`

But yes, this looks better.

**Comment by cynicaldevil** [comment]
@karroffel Updated PR with changes.

**Comment by karroffel** [comment]
@cynicaldevil uhm, so thanks very much for the initiative! But Image has been removed from the list of core types, so a wrapper isn't needed anymore. Actually most of the core API is complete now, the only thing that's really missing are string functions.

So it depends on you. I can close this and implement the string stuff myself or you can try to fill in the string stuff.

**Comment by cynicaldevil** [comment]
@karroffel I don't have time to contribute anymore, so I don't think I'll be able to write the string functions.

**Comment by karroffel** [comment]
@cynicaldevil that's okay, thanks for your initiative! 💙

**Pull request closed without merging**